File: apps/android/app/src/main/java/com/galaxyssi/chat/MqttDeliveryDispatch.kt
package com.galaxyssi.chat

import org.eclipse.paho.client.mqttv3.IMqttDeliveryToken
import org.eclipse.paho.client.mqttv3.MqttException
import java.util.Locale
import java.util.UUID

/** Bounded physical copies. The existing durable outbox, not this scheduler, owns restart/retry. */
internal class MqttDeliveryDispatch(
    private val policy: MqttMultipathPolicy,
    private val publish: (String, Long, String, ByteArray, String) -> Long?,
    private val completed: (IMqttDeliveryToken, Boolean) -> Unit,
    private val now: () -> Long
) : AutoCloseable {
    data class Delivery(val peer: String, val message: MqttDeliveryEnvelope.Message, val receiveTopics: Set<String>,
        val encodeAttempt: (MqttDeliveryEnvelope.Frame) -> ByteArray,
        val authorized: (String, Long) -> Boolean, val sizeBound: Int)
    private data class Scheduled(val due: Long, val plan: MqttMultipathPolicy.Dispatch)
    private class Job(val topic: String, val delivery: Delivery, val token: IMqttDeliveryToken,
        val started: Long, val scheduled: MutableList<Scheduled>) {
        val key = delivery.peer to delivery.message
        val attempts = mutableSetOf<String>()
        var accepted = false
        var publishing = false
        var finished = false
    }
    private data class Sent(val job: Job, val frame: MqttDeliveryEnvelope.Frame, var pending: Boolean = true)
    data class Diagnostics(val messages: Int, val trackedAttempts: Int, val queuedCopies: Int, val bufferedBytes: Long)
    private val lock = Any()
    private val jobs = linkedMapOf<Pair<String, MqttDeliveryEnvelope.Message>, Job>()
    private val sent = mutableMapOf<String, Sent>()
    private var closed = false
    private fun traffic(message: MqttDeliveryEnvelope.Message) = MqttMultipathPolicy.Traffic.valueOf(message.traffic.uppercase(Locale.ROOT))
    private fun priority(job: Job) = job.delivery.message.traffic in setOf("control", "receipt", "final")
    private fun disconnected(): Nothing = throw MqttException(MqttException.REASON_CODE_CLIENT_NOT_CONNECTED.toInt())

    fun submit(topic: String, delivery: Delivery, token: IMqttDeliveryToken): IMqttDeliveryToken {
        val at = now()
        val plans = policy.plan(delivery.peer, delivery.message.messageId, traffic(delivery.message),
            delivery.sizeBound, delivery.receiveTopics, at)
        val job: Job
        synchronized(lock) {
            val key = delivery.peer to delivery.message
            jobs[key]?.let { if (expired(it, at)) retire(it) }
            jobs[key]?.let { return it.token }
            if (closed || plans.isEmpty()) disconnected()
            val reserved = delivery.message.traffic !in setOf("control", "receipt", "final")
            if (jobs.size >= MqttBrokerCatalog.MAX_ATTEMPTS - (if (reserved) MqttBrokerCatalog.CONTROL_RESERVE else 0) ||
                jobs.values.sumOf { it.delivery.sizeBound.toLong() } + delivery.sizeBound > MqttBrokerCatalog.INFLIGHT_BYTES -
                (if (reserved) MqttBrokerCatalog.CONTROL_RESERVE * MqttBrokerCatalog.SMALL_PACKET_BYTES else 0)) {
                throw MqttException(MqttException.REASON_CODE_MAX_INFLIGHT.toInt())
            }
            job = Job(topic, delivery, token, at, plans.map { Scheduled(at + it.delayMs, it) }.toMutableList())
            jobs[key] = job
        }
        try { pump(job, at) } catch (error: Exception) {
            synchronized(lock) { if (job.attempts.isEmpty()) jobs.remove(job.key) }
            throw error
        }
        synchronized(lock) {
            if (job.attempts.isEmpty() && !job.accepted) {
                job.scheduled.clear()
                jobs.remove(job.key)
                disconnected()
            }
        }
        return token
    }

    private fun finish(job: Job, accepted: Boolean) {
        synchronized(lock) {
            if (job.finished) return
            job.finished = true
        }
        completed(job.token, accepted)
    }

    private fun pump(job: Job, at: Long) {
        synchronized(lock) {
            if (closed || job.accepted || job.publishing || jobs[job.key] !== job) return
            job.publishing = true
        }
        try {
            while (true) {
                val plan = synchronized(lock) {
                    if (closed || job.accepted || job.scheduled.isEmpty() || job.scheduled[0].due > at) return
                    job.scheduled.removeAt(0).plan
                }
                val delivery = job.delivery
                val message = delivery.message
                val allowed = policy.plan(delivery.peer, message.messageId, traffic(message), delivery.sizeBound,
                    delivery.receiveTopics, now())
                if (allowed.none { it.brokerId == plan.brokerId && it.generation == plan.generation } ||
                    !delivery.authorized(plan.brokerId, plan.generation)) continue
                val attemptId = UUID.randomUUID().toString().replace("-", "")
                val frame = MqttDeliveryEnvelope.Frame(message, MqttDeliveryEnvelope.Attempt(attemptId, plan.brokerId, plan.generation))
                val encoded = delivery.encodeAttempt(frame)
                val size = mqttPublishPacketBytes(job.topic, encoded.size)
                require(size <= delivery.sizeBound) { "Delivery exceeds final encoded packet bound" }
                val admitted = synchronized(lock) {
                    if (closed || job.accepted || jobs[job.key] !== job) return
                    if (!policy.reserve(attemptId, MqttMultipathPolicy.Attempt(delivery.peer, message.messageId,
                            message.contentHash, plan.brokerId, plan.generation, size.toInt(), traffic(message), now()))) {
                        job.scheduled.add(0, Scheduled(at + 250, plan))
                        false
                    } else {
                        job.attempts.add(attemptId)
                        sent[attemptId] = Sent(job, frame)
                        true
                    }
                }
                if (!admitted) break
                val result = runCatching { publish(plan.brokerId, plan.generation, job.topic, encoded, attemptId) }.getOrNull()
                if (result == null) synchronized(lock) {
                    sent.remove(attemptId)
                    job.attempts.remove(attemptId)
                    policy.discardAttempt(attemptId)
                    expedite(job, at)
                }
            }
        } catch (error: Exception) {
            synchronized(lock) { job.scheduled.clear() }
            throw error
        } finally {
            val failed = synchronized(lock) {
                job.publishing = false
                job.scheduled.isEmpty() && job.attempts.none { sent[it]?.pending == true }
            }
            if (failed) finish(job, false)
        }
    }

    private fun expedite(job: Job, at: Long) {
        for (index in job.scheduled.indices) job.scheduled[index] = job.scheduled[index].copy(due = at)
    }

    fun published(receipt: MqttBrokerPool.PublishReceipt): Boolean {
        val job: Job
        val failed: Boolean
        synchronized(lock) {
            val item = sent[receipt.attemptId] ?: return false
            val attempt = item.frame.attempt
            if (attempt.brokerId != receipt.physical.brokerId || attempt.generation != receipt.physical.generation) return true
            if (!item.pending) return true
            item.pending = false
            job = item.job
            if (receipt.brokerAcked) policy.brokerAck(attempt.attemptId, attempt.brokerId, attempt.generation)
            else {
                policy.discardAttempt(attempt.attemptId)
                sent.remove(attempt.attemptId)
                job.attempts.remove(attempt.attemptId)
                expedite(job, now())
            }
            if (job.accepted || closed) {
                sent.remove(attempt.attemptId)
                job.attempts.remove(attempt.attemptId)
                policy.discardAttempt(attempt.attemptId)
                if (job.attempts.isEmpty()) jobs.remove(job.key)
            }
            failed = !job.publishing && job.scheduled.isEmpty() && job.attempts.none { sent[it]?.pending == true }
        }
        if (receipt.brokerAcked) finish(job, true) else if (failed) finish(job, false)
        return true
    }

    /** Only after current pair AEAD validation. A failed durable commit leaves retries live. */
    fun acceptVerifiedReceipt(peer: String, frame: MqttDeliveryEnvelope.Frame, commit: () -> Unit): Boolean {
        val job = synchronized(lock) {
            val item = sent[frame.attempt.attemptId]
            if (closed || item == null || item.frame != frame || item.job.delivery.peer != peer || item.job.accepted) return false
            item.job
        }
        commit()
        synchronized(lock) {
            val accepted = policy.acceptVerifiedReceipt(peer, frame.message.messageId, frame.message.contentHash,
                frame.attempt.attemptId, now())
            if (accepted.isEmpty()) return false
            job.accepted = true
            job.scheduled.clear()
            accepted.forEach { key ->
                sent[key]?.takeIf { !it.pending }?.let { sent.remove(key); it.job.attempts.remove(key) }
            }
            return true
        }
    }

    /** Stored-message ACKs without attempt attribution cannot supply path RTT samples. */
    fun acceptVerifiedMessage(peer: String, messageId: String, contentHash: String) = synchronized(lock) {
        if (closed) return@synchronized
        policy.acceptVerifiedMessage(peer, messageId, contentHash)
        jobs.values.filter { it.delivery.peer == peer && it.delivery.message.messageId == messageId &&
            it.delivery.message.contentHash == contentHash }.forEach { it.accepted = true; it.scheduled.clear() }
    }

    fun tick() {
        val at = now()
        val selected = synchronized(lock) {
            val due = jobs.values.filter { expired(it, at) || it.accepted || it.scheduled.firstOrNull()?.due?.let { due -> due <= at } == true }
            val (urgent, ordinary) = due.partition(::priority)
            val batch = (urgent.take(12) + ordinary.take(4)).toMutableList()
            batch.addAll((urgent.drop(12) + ordinary.drop(4)).take(16 - batch.size))
            batch.forEach { jobs.remove(it.key); jobs[it.key] = it }
            batch
        }
        for (job in selected) {
            val retired = synchronized(lock) {
                if (expired(job, at) || job.accepted) { retire(job); true } else false
            }
            if (!retired) pump(job, at)
        }
    }

    private fun expired(job: Job, at: Long) = at - job.started >= MqttBrokerCatalog.ATTEMPT_OBSERVATION_MS
    private fun retire(job: Job) {
        if (job.publishing) return
        job.scheduled.clear()
        job.attempts.toList().forEach { key ->
            sent[key]?.takeIf { !it.pending }?.let { sent.remove(key); policy.discardAttempt(key); job.attempts.remove(key) }
        }
        if (job.attempts.isEmpty()) jobs.remove(job.key)
    }

    override fun close() {
        val pending = synchronized(lock) { closed = true; jobs.values.toList().also { it.forEach(::retire) } }
        pending.forEach { finish(it, false) }
    }

    fun diagnostics() = synchronized(lock) {
        Diagnostics(jobs.size, sent.size, jobs.values.sumOf { it.scheduled.size }, jobs.values.sumOf { it.delivery.sizeBound.toLong() })
    }
}

File: apps/desktop/core/galaxyssi-link/backend/mqtt_delivery_dispatch.py
"""Bounded physical races over the existing durable Link outbox.

No worker or durable message queue lives here. The pool owner's maintenance
tick drives delayed copies; only authenticated RX_STORED ends their race.
"""
from __future__ import annotations

from dataclasses import dataclass, field
import secrets
import threading
import time
from typing import Callable

import paho.mqtt.client as mqtt

from mqtt_broker_pool import publish_packet_bytes
from mqtt_broker_catalog import CATALOG
from mqtt_delivery_envelope import Attempt as WireAttempt, Frame, Message
from mqtt_multipath_policy import Attempt, Traffic


@dataclass(frozen=True)
class Delivery:
    peer: str
    message: Message
    receive_topics: frozenset[str]
    encode_attempt: Callable[[Frame], bytes | str]
    authorized: Callable[[str, int], bool]
    size_bound: int


@dataclass
class _Job:
    topic: str
    delivery: Delivery
    info: object
    started: float
    scheduled: list
    attempts: set = field(default_factory=set)
    accepted: bool = False
    publishing: bool = False

    @property
    def key(self):
        return self.delivery.peer, self.delivery.message


@dataclass
class _Sent:
    job: _Job
    frame: Frame
    started: float
    pending: bool = True


class DeliveryDispatch:
    def __init__(self, policy, publish, completed, *, clock=time.monotonic):
        self.policy, self._publish, self._completed, self._clock = policy, publish, completed, clock
        self._jobs, self._sent = {}, {}
        self._lock = threading.RLock()
        self._closed = False

    def submit(self, topic, delivery, info):
        now = self._clock()
        traffic = Traffic(delivery.message.traffic)
        plans = self.policy.plan(delivery.peer, delivery.message.message_id, traffic,
                                 delivery.size_bound, set(delivery.receive_topics), now=now)
        with self._lock:
            key = delivery.peer, delivery.message
            previous = self._jobs.get(key)
            if previous is not None and self._expired(previous, now):
                self._retire(previous)
                previous = self._jobs.get(key)
            if previous is not None:
                return previous.info
            priority = traffic in {Traffic.CONTROL, Traffic.RECEIPT, Traffic.FINAL}
            limit = self.policy.limits
            if self._closed or not plans:
                info.rc = mqtt.MQTT_ERR_NO_CONN
                return info
            if (len(self._jobs) >= limit.max_attempts - (0 if priority else limit.control_reserve)
                    or sum(job.delivery.size_bound for job in self._jobs.values()) + delivery.size_bound
                    > limit.inflight_bytes - (0 if priority else limit.control_reserve * limit.small_packet_bytes)):
                info.rc = mqtt.MQTT_ERR_QUEUE_SIZE
                return info
            job = _Job(topic, delivery, info, now, [(now + plan.delay, plan) for plan in plans])
            self._jobs[key] = job
        try:
            self._pump(job, now)
        except Exception:
            with self._lock:
                if not job.attempts:
                    self._jobs.pop(key, None)
            raise
        with self._lock:
            # Admission requires a physical send, not another unbounded queue.
            if not job.attempts:
                job.scheduled.clear()
                self._jobs.pop(key, None)
                if not info._done.is_set():
                    info.rc = mqtt.MQTT_ERR_QUEUE_SIZE
        return info

    def _finish(self, job, success):
        with self._lock:
            if job.info._done.is_set():
                return
            job.info.rc = mqtt.MQTT_ERR_SUCCESS if success else mqtt.MQTT_ERR_NO_CONN
            job.info._done.set()
        self._completed(job.info, success)

    def _pump(self, job, now):
        with self._lock:
            if self._closed or job.accepted or job.publishing or self._jobs.get(job.key) is not job:
                return
            job.publishing = True
        try:
            while True:
                with self._lock:
                    if self._closed or job.accepted or not job.scheduled or job.scheduled[0][0] > now:
                        break
                    _due, plan = job.scheduled.pop(0)
                delivery, message = job.delivery, job.delivery.message
                # Revalidate expiry, subscriptions, generation AND pair identity
                # at each actual send, not just when the outbox selected a row.
                allowed = self.policy.plan(delivery.peer, message.message_id, Traffic(message.traffic),
                                           delivery.size_bound, set(delivery.receive_topics), now=now)
                if (not any((p.broker_id, p.generation) == (plan.broker_id, plan.generation) for p in allowed)
                        or not delivery.authorized(plan.broker_id, plan.generation)):
                    continue
                frame = Frame(message, WireAttempt(secrets.token_hex(16), plan.broker_id, plan.generation))
                encoded = delivery.encode_attempt(frame)
                encoded = encoded.encode("utf-8") if isinstance(encoded, str) else bytes(encoded)
                size = publish_packet_bytes(job.topic, len(encoded))
                if size > delivery.size_bound:
                    raise ValueError("Delivery exceeds its encoded packet bound")
                with self._lock:
                    if self._closed or job.accepted or self._jobs.get(job.key) is not job:
                        break
                    attempt_id = frame.attempt.attempt_id
                    if not self.policy.reserve(attempt_id, Attempt(delivery.peer, message.message_id,
                            message.content_hash, plan.broker_id, plan.generation, size, Traffic(message.traffic), now)):
                        job.scheduled.insert(0, (now + 0.25, plan))
                        break
                    job.attempts.add(attempt_id)
                    self._sent[attempt_id] = _Sent(job, frame, now)
                try:
                    receipt = self._publish(plan.broker_id, plan.generation, job.topic, encoded, attempt_id=attempt_id)
                except Exception:
                    receipt = None
                if receipt is None:
                    with self._lock:
                        self._sent.pop(attempt_id, None)
                        job.attempts.discard(attempt_id)
                        self.policy.discard_attempt(attempt_id)
                        job.scheduled = [(now, p) for _, p in job.scheduled]
        except Exception:
            with self._lock:
                job.scheduled.clear()
            raise
        finally:
            with self._lock:
                job.publishing = False
                failed = not job.scheduled and not any(
                    sent.pending for key in job.attempts if (sent := self._sent.get(key)))
            if failed and not job.info._done.is_set():
                self._finish(job, False)

    def published(self, receipt):
        with self._lock:
            sent = self._sent.get(receipt.attempt_id)
            if sent is None:
                return False
            attempt = sent.frame.attempt
            if (attempt.broker_id, attempt.generation) != (receipt.physical.broker_id, receipt.physical.generation):
                return True
            if not sent.pending:
                return True
            sent.pending = False
            job = sent.job
            if receipt.broker_acked:
                self.policy.broker_ack(attempt.attempt_id, attempt.broker_id, attempt.generation)
            else:
                self.policy.discard_attempt(attempt.attempt_id)
                self._sent.pop(attempt.attempt_id, None)
                job.attempts.discard(attempt.attempt_id)
                job.scheduled = [(self._clock(), plan) for _, plan in job.scheduled]
            if job.accepted or self._closed:
                self._sent.pop(attempt.attempt_id, None)
                job.attempts.discard(attempt.attempt_id)
                self.policy.discard_attempt(attempt.attempt_id)
                if not job.attempts:
                    self._jobs.pop(job.key, None)
            failed = not job.publishing and not job.scheduled and not any(
                item.pending for key in job.attempts if (item := self._sent.get(key)))
        if receipt.broker_acked:
            self._finish(job, True)
        elif failed:
            self._finish(job, False)
        return True

    def accept_verified_receipt(self, peer, frame, commit):
        """After current-pair AEAD validation, bind to our exact local attempt.

        commit may retire the existing durable outbox row. Its failure must
        leave transport retries live. There is no receipt-of-receipt here.
        """
        with self._lock:
            sent = self._sent.get(frame.attempt.attempt_id)
            if self._closed or not sent or sent.frame != frame or sent.job.delivery.peer != peer or sent.job.accepted:
                return False
            job = sent.job
        commit()
        with self._lock:
            completed = self.policy.accept_verified_receipt(peer, frame.message.message_id,
                frame.message.content_hash, frame.attempt.attempt_id, now=self._clock())
            if not completed:
                return False
            job.accepted = True
            job.scheduled.clear()
            # Sent copies stay accounted until PUBACK/disconnect. Unsent copies
            # are cancelled; a business receipt is not a fake broker receipt.
            for key in completed:
                other = self._sent.get(key)
                if other and not other.pending:
                    self._sent.pop(key, None)
                    other.job.attempts.discard(key)
            return True

    def accept_verified_message(self, peer, message_id, content_hash):
        """Called only after the existing ciphertext-bound outbox ACK passes."""
        with self._lock:
            if self._closed:
                return
            self.policy.accept_verified_message(peer, message_id, content_hash)
            for job in self._jobs.values():
                message = job.delivery.message
                if (job.delivery.peer, message.message_id, message.content_hash) == (peer, message_id, content_hash):
                    job.accepted = True
                    job.scheduled.clear()

    def tick(self):
        now = self._clock()
        with self._lock:
            due = [job for job in self._jobs.values() if self._expired(job, now) or job.accepted
                   or (job.scheduled and job.scheduled[0][0] <= now)]
            urgent = [job for job in due if job.delivery.message.traffic in {"control", "final", "receipt"}]
            ordinary = [job for job in due if job.delivery.message.traffic not in {"control", "final", "receipt"}]
            jobs = urgent[:12] + ordinary[:4]
            jobs += (urgent[12:] + ordinary[4:])[:16 - len(jobs)]
            for job in jobs:
                self._jobs.pop(job.key, None)
                self._jobs[job.key] = job
        for job in jobs:
            with self._lock:
                if self._expired(job, now) or job.accepted:
                    self._retire(job)
                    continue
            self._pump(job, now)

    @staticmethod
    def _expired(job, now):
        return now - job.started >= CATALOG["timing"]["attempt_observation_seconds"]

    def _retire(self, job):
        if job.publishing:
            return
        job.scheduled.clear()
        for key in list(job.attempts):
            sent = self._sent.get(key)
            if sent and not sent.pending:
                self._sent.pop(key, None)
                self.policy.discard_attempt(key)
                job.attempts.discard(key)
        if not job.attempts:
            self._jobs.pop(job.key, None)

    def close(self):
        with self._lock:
            self._closed = True
            jobs = list(self._jobs.values())
            for job in jobs:
                self._retire(job)
        for job in jobs:
            self._finish(job, False)

    def diagnostics(self):
        with self._lock:
            return {"messages": len(self._jobs), "tracked_attempts": len(self._sent),
                    "queued_copies": sum(len(job.scheduled) for job in self._jobs.values()),
                    "buffered_bytes": sum(job.delivery.size_bound for job in self._jobs.values())}
